### PR TITLE
Add 3.12 typing features to the compiler

### PR DESCRIPTION
**Pull request by dchiquito**
This is the first step towards implementing 3.12 compliant typing(#5252 ) . It's taken me a long time to get this far so I wanted to check in before spending more time building on this foundation.

Changes made:
* Add the builtin `_typing` stdlib module. This will eventually replace the fundamental typing primitives that are currently implemented in `typing.py`. This is still very much a work in progress, but it works well enough to exercise the new compiler features.
* Compile and evaluate function generics (`def foo[X](x: X): ...`)
* Compile and evaluate class generics (`class C[X]: def __init__(x: X): ...`)
* Compile type alias statements (`type X = list[int]`)

Some questions:
* The `push_symbol_table` and `pop_symbol_table` are my attempt to simulate the behavior of the new [Annotation Scopes](https://docs.python.org/3/reference/executionmodel.html#annotation-scopes). I'm not sure if I got this 100% right.
* At some point, CPython introduced the compiler-level concept of [intrinsics](https://github.com/python/cpython/blob/v3.12.2/Python/intrinsics.c#L212). There are a handful of instructions like `CALL_INTRINSIC_1` and `CALL_INTRINSIC_2` that do weird fundamental python things like printing, importing, or defining types. It seems to me that this is mostly a way of organizing the code than a fundamental design requirement, so I ignored this and added the intrinsics I needed as instructions. Perhaps this is not correct?
* I made a small change to how the bookkeeping is done for unmarshalling bytecode. Adding a new instruction caused me some very odd intermittent errors that took me a while to debug, so my hope is to prevent headaches in the future.

### CODE REVIEW

File: vm/src/frame.rs [hunk 1]
```diff
@@ -1659,6 +1699,14 @@ impl ExecutingFrame<'_> {
             vm.ctx.new_dict().into()
         };
 
+        let type_params: PyTupleRef = if flags.contains(bytecode::MakeFunctionFlags::TYPE_PARAMS) {
+            self.pop_value()
+                .downcast()
+                .map_err(|_| vm.new_type_error("Type params must be a tuple.".to_owned()))?
+        } else {
+            vm.ctx.new_tuple(vec![])
```
Comment (youknowone) [inline vm/src/frame.rs:1707]:
```suggestion
            vm.ctx.empty_tuple.clone()
```
empty_tuple shares a singleton object

**Review by youknowone — COMMENTED** [review]
Awesome! I like the changes.

> The push_symbol_table and pop_symbol_table are my attempt to simulate the behavior of the new [Annotation Scopes](https://docs.python.org/3/reference/executionmodel.html#annotation-scopes). I'm not sure if I got this 100% right.

To be honest, I don't know well about either Annotation Scopes or our symbol table. If you are not sure about the design, leaving a comment with your concern will be fine.
@coolreader18 Do you have any advice?

> At some point, CPython introduced the compiler-level concept of [intrinsics](https://github.com/python/cpython/blob/v3.12.2/Python/intrinsics.c#L212). There are a handful of instructions like CALL_INTRINSIC_1 and CALL_INTRINSIC_2 that do weird fundamental python things like printing, importing, or defining types. It seems to me that this is mostly a way of organizing the code than a fundamental design requirement, so I ignored this and added the intrinsics I needed as instructions. Perhaps this is not correct?

Yes, CPython is recently very quickly changing its instruction design. Our instructions design is staying around 3.6~3.8. Your approach seems correct and fits in traditional Python instruction design well. We will be able to introduce those new concepts later. Don't worry!

> I made a small change to how the bookkeeping is done for unmarshalling bytecode. Adding a new instruction caused me some very odd intermittent errors that took me a while to debug, so my hope is to prevent headaches in the future.

I am sorry to make you to be confused. How about placing a comment to the last instruction like "Please add instruction in the middle of the enum; otherwise `LAST_INSTRUCTION` need to be updated"

**Comment by dchiquito** [comment]
Thanks for the feedback! Would you prefer keeping this branch separate until typing is finished or merging it now?

**Comment by youknowone** [comment]
I prefer to merge small patches, frequently. I will review this patch soon, thank you so much!

**Review by youknowone — APPROVED** [review]
Awesome, finally we have `_typing`!

I also tried #4946 on it. `_typing` look like almost close to working.

Thank you for making a big progress 😄

**Pull request merged**